x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -159,14 +178,45 @@ private <T> void preAnalyze(LogicalPlan parsed, BiFunction<IndexResolution, Enri
             l.onResponse(action.apply(resolvedIndex.get(), resolution));
         });
         try (RefCountingListener refs = new RefCountingListener(groupedListener)) {
-            preAnalyzeIndices(parsed, refs.acquire(resolvedIndex::set));
+            preAnalyzeIndices(parsed, refs.acquire(resolvedIndex::set), fieldNames);
             for (String policyName : policyNames) {
                 enrichPolicyResolver.resolvePolicy(policyName, refs.acquire(resolution.resolvedPolicies()::add));
             }
         }
     }
 
-    private <T> void preAnalyzeIndices(LogicalPlan parsed, ActionListener<IndexResolution> listener) {
+    private <T> void preAnalyzeSequentially(
+        LogicalPlan parsed,
+        BiFunction<IndexResolution, EnrichResolution, T> action,
+        Set<String> policyNames,
+        EnrichResolution enrichResolution,
+        Set<String> fieldNames,
+        ActionListener<T> listener
+    ) {
+        ActionListener<Void> groupedListener = listener.delegateFailureAndWrap((l, unused) -> {
+            assert enrichResolution.resolvedPolicies().size() == policyNames.size()
+                : enrichResolution.resolvedPolicies().size() + " != " + policyNames.size();
+            var matchFields = enrichResolution.resolvedPolicies()
+                .stream()
+                .filter(p -> p.index().isValid())
```

Comment:
The enrich policy maybe cannot be found, thus the check here. But not doing anything here, so that the Verifier catches the missing policy and report it appropriately.